text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1681 @@
+
+
+### API Docs
+
+There is presently no API for `<template>` itself, as proposed in this RFC, though it leaves room for future RFCs to do so.[^emblem-etc] Since there is no import location, we should cover it under the `@glimmer/component` module documentation. This will be a natural home for it, since we will always discuss it in the context of components.
+
+[^emblem-etc]: Historically, for example, many Ember apps used [Emblem][emblem] as a templating language—and it is still possible to do so today! In the future, that could be supported with `<template lang="emblem">`. This would also be an easy home for experiments with a Svelte-like syntax with e.g. `<template lang="svelte">` etc.
+
+[emblem]: http://emblemjs.com
+
+
+### Existing Ember users
+
+The Ember community has long experience with the idea that we can only have *one* component per file, and that component templates and the backing class must always be in separate files. The name of this feature, *first-class component templates*, is designed to help explain how it relates to that historical experience. In the past, templates in many ways were second-class citizens of the overall experience of authoring an Ember app—especially for template-only components. Adding even a small amount of functionality to a template came with a lot of friction and other downsides: switching to a class-backed component, or introducing a globally-available helper. Now, adding a local helper utility is no different for a template-only component than it is anywhere else: write a function!
+
+As a result, we can teach this to existing Ember users as taking everything they already know about how templates work, while making it faster, easier, and lighter-weight to solve common problems. The one significant new concept here is the `<template>` tag. We can describe that to existing features as the one addition which raises templates up to being a first-class tool in your app or addon. For this audience, we can explicitly connect standalone `<template>` to “template-only components,” a long-standing concept. (I did *not* use this verbiage in the overall discussion because this design, even more than Octane, helps make template-only components the *default* kind of component, with class-backed components the exception and thing to treat as a special case.)
+
+A blog post can introduce the feature along these lines when the feature ships, with examples showing how it simplifies existing patterns and enables new capabilities. For simplifying existing patterns, we might pull the same Mapbox example from the guides shown above. For enabling new capabilities, we could show how it enables using native CSS Modules with none of the hacks required by Ember CSS Modules.
+
+
+## Drawbacks
```

Comment:
IMHO a drawback that should be listed is that introducing a new file extension will lack editor support for a very long time (if not forever) on uncommon editors. I personally think it is ok to have this, but I also think we should be very aware of it.

Reply:
imo, most editors already don't support what we need for ember.
Ime, maintainers of more obscure editor/tool plugins tend to stick to the vanilla handlebars and don't let us add anything specific to ember. 

So, imo, having _any_ ember specific plugin would be an improvement.

Reply:
True, but my point is: having *any* hbs syntax support is *much* better then nothing. And if its only highlighting `{{`. Also currently this is only a problem for templates, but this would make it a problem for *all* code, also `.js`. People may stop having any `.js` files, in case they wanna add a `<template>`, and so we move from a *good js support and bad hbs support* on exotic editors to *no support for any file at all*.

Reply:
I mean, we can declare certain levels of editor support before shipping the feature -- I don't think that should block the exploration and implementation of the `<template>` feature.

For example, I think the following are required before gjs/gts:
- VSCode
  - syntax fully supported
- (neo)vim
  - syntax partially supported via treesitter
- Webstorm / IntelliJ
  - needs exploration
- Atom (also includes GitHub highlighting)
  - in-progress: https://github.com/atom/language-javascript/pull/710 (very early -- need to revisit this -- I don't use Atom so this has been slightly low priority)

Non-Editor tools
- eslint
  - parsing & identifier usage: in-progress: https://github.com/ember-cli/eslint-plugin-ember/pull/1395
- template-lint:
  - parsing & identifier usage: in-progress: https://github.com/ember-template-lint/ember-template-lint/pull/2324
- highlight.js
 - full support via glimmer plugin: https://hljs-glimmer.nullvoxpopuli.com/?sample=gjs
- prism.js:
 - full support via glimmer plugin: https://prismjs-glimmer.nullvoxpopuli.com/?sample=gjs

These are stretch goals 
 - Monaco 
 - CodeMirror
 
 (I use both of these on https://limber.glimdown.com, so "eventually" we'll get gjs/gts support _in_ markdown as well :tada: )
 
 For each editor's lint plugins' they'll get support for `<template>` via the eslint-plugin-ember and template-lint PR changes.  
 the unstable-ember-template-server and any editor that has an integration with that would likely be the first to have in-editor template-linting for `<template>` (need to sync with @lifeart) on this.